File: scripts/import-excel-data.mjs
import openpyxl from 'openpyxl';
import { getDb } from '../server/db.ts';
import { errorReports } from '../drizzle/schema.ts';

/**
 * Script para importar dados do Excel para o banco de dados
 * Uso: node scripts/import-excel-data.mjs <caminho-do-arquivo>
 */

const excelFilePath = process.argv[2];

if (!excelFilePath) {
  console.error('❌ Erro: Caminho do arquivo Excel não fornecido');
  console.error('Uso: node scripts/import-excel-data.mjs <caminho-do-arquivo>');
  process.exit(1);
}

async function importExcelData() {
  try {
    console.log('📂 Carregando arquivo Excel...');
    
    // Usar Python para ler o Excel (já que openpyxl não está disponível em Node)
    const { execSync } = await import('child_process');
    const fs = await import('fs');
    
    // Criar script Python temporário
    const pythonScript = `
import openpyxl
import json
import sys

filename = sys.argv[1]
wb = openpyxl.load_workbook(filename)
ws = wb.active

data = []
for i, row in enumerate(ws.iter_rows(min_row=4, values_only=True)):
    if row[0]:  # Se tem ID
        data.append({
            'id': row[0],
            'key': str(int(row[2])) if row[2] else None,
            'modules': row[3] or 'Não especificado',
            'origin': row[4] or 'Outro',
            'reason': row[5] or 'Em análise',
            'agent': row[6] or 'Não atribuído',
            'records': row[7] or '',
            'status': row[8] or 'No prazo',
            'ticket': row[9] or '',
            'action': row[10] or ''
        })

print(json.dumps(data, ensure_ascii=False))
`;

    const tempPythonFile = '/tmp/import_excel.py';
    fs.writeFileSync(tempPythonFile, pythonScript);
    
    console.log('🔄 Lendo dados do Excel...');
    const result = execSync(`python3 ${tempPythonFile} "${excelFilePath}"`).toString();
    const data = JSON.parse(result);
    
    console.log(`✅ Encontrados ${data.length} reports para importar`);
    
    // Conectar ao banco de dados
    const db = await getDb();
    if (!db) {
      throw new Error('Não foi possível conectar ao banco de dados');
    }
    
    // Mapear valores para os enums corretos
    const mapOrigin = (origin) => {
      const map = {
        'Onbording': 'onboarding',
        'Onboarding': 'onboarding',
        'Outro': 'other'
      };
      return map[origin] || 'other';
    };
    
    const mapReason = (reason) => {
      const map = {
        'Cliente (Base)': 'client_base',
        'Cliente Base': 'client_base',
        'Modelador': 'modeler',
        'Analista': 'analyst',
        'Engenharia': 'engineering',
        'Em análise': 'in_analysis'
      };
      return map[reason] || 'in_analysis';
    };
    
    const mapStatus = (status) => {
      const map = {
        'No prazo': 'on_time',
        'SLA Vencida': 'sla_expired',
        'Crítico': 'critical',
        'Resolvido': 'resolved'
      };
      return map[status] || 'on_time';
    };
    
    // Importar dados
    let imported = 0;
    let errors = 0;
    
    for (const report of data) {
      try {
        await db.insert(errorReports).values({
          clientId: report.id,
          key: report.key,
          modules: report.modules,
          origin: mapOrigin(report.origin),
          reason: mapReason(report.reason),
          agent: report.agent,
          records: report.records,
          status: mapStatus(report.status),
          ticket: report.ticket || null,
          recommendedAction: report.action || null,
          createdAt: new Date(),
          updatedAt: new Date()
        });
        imported++;
        
        if (imported % 10 === 0) {
          console.log(`  ✓ ${imported} reports importados...`);
        }
      } catch (err) {
        errors++;
        console.error(`  ❌ Erro ao importar report ${report.id}:`, err.message);
      }
    }
    
    console.log(`\n✅ Importação concluída!`);
    console.log(`   📊 ${imported} reports importados com sucesso`);
    if (errors > 0) {
      console.log(`   ⚠️  ${errors} erros durante a importação`);
    }
    
    // Limpar arquivo temporário
    fs.unlinkSync(tempPythonFile);
    
    process.exit(0);
  } catch (error) {
    console.error('❌ Erro durante importação:', error.message);
    process.exit(1);
  }
}

importExcelData();

File: scripts/import_excel_data.py
#!/usr/bin/env python3
"""
Script para importar dados do Excel para o banco de dados N0 Error Tracker
Uso: python3 scripts/import_excel_data.py <caminho-do-arquivo> [database-url]
"""

import openpyxl
import sys
import os
from datetime import datetime
import mysql.connector

def map_origin(origin):
    """Mapear origem para enum do banco"""
    mapping = {
        'Onbording': 'onboarding',
        'Onboarding': 'onboarding',
        'Outro': 'other'
    }
    return mapping.get(origin, 'other')

def map_reason(reason):
    """Mapear motivo para enum do banco"""
    mapping = {
        'Cliente (Base)': 'client_base',
        'Cliente Base': 'client_base',
        'Modelador': 'modeler',
        'Analista': 'analyst',
        'Engenharia': 'engineering',
        'Em análise': 'in_analysis'
    }
    return mapping.get(reason, 'in_analysis')

def map_status(status):
    """Mapear status para enum do banco"""
    mapping = {
        'No prazo': 'on_time',
        'SLA Vencida': 'sla_expired',
        'Crítico': 'critical',
        'Resolvido': 'resolved'
    }
    return mapping.get(status, 'on_time')

def import_excel_data(excel_file, db_url):
    """Importar dados do Excel para o banco de dados"""
    
    print(f"📂 Carregando arquivo Excel: {excel_file}")
    
    # Validar arquivo
    if not os.path.exists(excel_file):
        print(f"❌ Erro: Arquivo não encontrado: {excel_file}")
        sys.exit(1)
    
    # Carregar workbook
    try:
        wb = openpyxl.load_workbook(excel_file)
        ws = wb.active
    except Exception as e:
        print(f"❌ Erro ao abrir arquivo Excel: {e}")
        sys.exit(1)
    
    # Extrair dados (começando da linha 4, pulando headers)
    data = []
    for i, row in enumerate(ws.iter_rows(min_row=4, values_only=True), 1):
        if row[0]:  # Se tem ID
            try:
                # Validar que a chave é um número
                if not row[2]:
                    continue
                try:
                    key_int = int(float(row[2]))
                except (ValueError, TypeError):
                    continue
                
                data.append({
                    'id': str(row[0]).strip(),
                    'key': str(key_int),
                    'modules': str(row[3]).strip() if row[3] else 'Não especificado',
                    'origin': str(row[4]).strip() if row[4] else 'Outro',
                    'reason': str(row[5]).strip() if row[5] else 'Em análise',
                    'agent': str(row[6]).strip() if row[6] else 'Não atribuído',
                    'records': str(row[7]).strip() if row[7] else '',
                    'status': str(row[8]).strip() if row[8] else 'No prazo',
                    'ticket': str(row[9]).strip() if row[9] else '',
                    'action': str(row[10]).strip() if row[10] else ''
                })
            except Exception as e:
                print(f"⚠️  Erro ao processar linha {i}: {e}")
                continue
    
    print(f"✅ Encontrados {len(data)} reports para importar")
    
    # Conectar ao banco de dados
    print("🔄 Conectando ao banco de dados...")
    try:
        # Parse database URL
        # Formato: mysql://user:password@host:port/database
        db_url = db_url.replace('mysql://', '')
        # Dividir em user:password@host:port/database
        auth_part, host_part = db_url.split('@')
        user, password = auth_part.split(':')
        host_db_part, database = host_part.split('/')
        # Separar host e port
        if ':' in host_db_part:
            host, port = host_db_part.split(':')
            port = int(port)
        else:
            host = host_db_part
            port = 3306
        
        conn = mysql.connector.connect(
            host=host,
            user=user,
            password=password,
            database=database,
            port=port
        )
        cursor = conn.cursor()
    except Exception as e:
        print(f"❌ Erro ao conectar ao banco de dados: {e}")
        sys.exit(1)
    
    # Importar dados
    imported = 0
    errors = 0
    
    for report in data:
        try:
            query = """
            INSERT INTO error_reports 
            (clientId, key, modules, origin, reason, agent, records, status, ticket, recommendedAction, createdAt, updatedAt)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            
            values = (
                report['id'],
                report['key'],
                report['modules'],
                map_origin(report['origin']),
                map_reason(report['reason']),
                report['agent'],
                report['records'],
                map_status(report['status']),
                report['ticket'] if report['ticket'] else None,
                report['action'] if report['action'] else None,
                datetime.now(),
                datetime.now()
            )
            
            cursor.execute(query, values)
            imported += 1
            
            if imported % 10 == 0:
                print(f"  ✓ {imported} reports importados...")
        
        except Exception as e:
            errors += 1
            print(f"  ❌ Erro ao importar report {report['id']}: {e}")
    
    # Confirmar transação
    try:
        conn.commit()
        cursor.close()
        conn.close()
    except Exception as e:
        print(f"❌ Erro ao confirmar importação: {e}")
        sys.exit(1)
    
    print(f"\n✅ Importação concluída!")
    print(f"   📊 {imported} reports importados com sucesso")
    if errors > 0:
        print(f"   ⚠️  {errors} erros durante a importação")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("❌ Erro: Caminho do arquivo Excel não fornecido")
        print("Uso: python3 scripts/import_excel_data.py <caminho-do-arquivo> [database-url]")
        sys.exit(1)
    
    excel_file = sys.argv[1]
    
    # Usar DATABASE_URL da variável de ambiente se não fornecido
    db_url = sys.argv[2] if len(sys.argv) > 2 else os.environ.get('DATABASE_URL')
    
    if not db_url:
        print("❌ Erro: DATABASE_URL não fornecido e não encontrado nas variáveis de ambiente")
        sys.exit(1)
    
    import_excel_data(excel_file, db_url)
